File: agapay-web/src/lib/monitoring.ts
import type { AlertTier } from "@/data/mockData";

export type MonitoringSample = {
  sequence_no: number;
  water_depth_cm: number | null;
  rainfall_mm: number;
  sensor_quality: "valid" | "invalid";
  device_uptime_ms: number;
  recorded_at: string;
};

export type MonitoringStation = {
  station_id: string;
  station_name: string;
  barangay: string | null;
  municipality: string | null;
  latitude: number | null;
  longitude: number | null;
  administrative_status: "active" | "inactive" | "maintenance";
  connection_status: "online" | "offline" | "no_data" | "inactive" | "maintenance";
  is_online: boolean;
  is_stale: boolean;
  age_seconds: number | null;
  alert_tier: Exclude<AlertTier, "OFFLINE" | "UNKNOWN"> | null;
  trend: "rising_rapidly" | "rising" | "stable" | "falling";
  current_depth_cm: number | null;
  latest_depth_cm: number | null;
  latest_rainfall_mm: number | null;
  sensor_quality: "valid" | "invalid" | null;
  last_ping: string | null;
  observed_at: string | null;
  firmware_version: string | null;
  source: "simulator" | "device" | "none";
  threshold_advisory_cm: number;
  threshold_warning_cm: number;
  threshold_evacuate_cm: number;
  history: MonitoringSample[];
};

export type MonitoringSnapshot = {
  checked_at: string;
  stale_after_seconds: number;
  stations: MonitoringStation[];
};

export function ageLabel(seconds: number | null, connection: MonitoringStation["connection_status"]) {
  if (seconds === null) return connection === "no_data" ? "No readings yet" : "Unknown";
  if (seconds < 60) return `${seconds} sec ago`;
  const minutes = Math.floor(seconds / 60);
  if (minutes < 60) return `${minutes} min ago`;
  const hours = Math.floor(minutes / 60);
  return `${hours} hr${hours === 1 ? "" : "s"} ago`;
}

File: agapay-mobile/lib/models/monitoring.dart
import 'alert_level.dart';

class MonitoringSample {
  MonitoringSample.fromJson(Map<String, dynamic> json)
    : sequenceNo = json['sequence_no'] as int,
      waterDepthCm = (json['water_depth_cm'] as num?)?.toDouble(),
      rainfallMm = (json['rainfall_mm'] as num).toDouble(),
      sensorQuality = json['sensor_quality'] as String,
      recordedAt = DateTime.parse(json['recorded_at'] as String).toLocal();

  final int sequenceNo;
  final double? waterDepthCm;
  final double rainfallMm;
  final String sensorQuality;
  final DateTime recordedAt;
}

class MonitoringStation {
  MonitoringStation.fromJson(Map<String, dynamic> json)
    : id = json['station_id'] as String,
      name = json['station_name'] as String,
      barangay = (json['barangay'] as String?) ?? 'Barangay unavailable',
      municipality = (json['municipality'] as String?) ?? 'Urdaneta City',
      latitude = (json['latitude'] as num?)?.toDouble(),
      longitude = (json['longitude'] as num?)?.toDouble(),
      connectionStatus = json['connection_status'] as String,
      isOnline = json['is_online'] as bool,
      isStale = json['is_stale'] as bool,
      ageSeconds = json['age_seconds'] as int?,
      alertLevel = _alert(json['alert_tier'] as String?),
      trend = json['trend'] as String,
      currentDepthCm = (json['current_depth_cm'] as num?)?.toDouble(),
      latestDepthCm = (json['latest_depth_cm'] as num?)?.toDouble(),
      latestRainfallMm = (json['latest_rainfall_mm'] as num?)?.toDouble(),
      sensorQuality = json['sensor_quality'] as String?,
      observedAt = _time(json['observed_at']),
      firmwareVersion = json['firmware_version'] as String?,
      source = json['source'] as String,
      thresholdAdvisoryCm = (json['threshold_advisory_cm'] as num).toDouble(),
      thresholdWarningCm = (json['threshold_warning_cm'] as num).toDouble(),
      thresholdEvacuateCm = (json['threshold_evacuate_cm'] as num).toDouble(),
      history = (json['history'] as List<dynamic>)
          .map(
            (value) => MonitoringSample.fromJson(value as Map<String, dynamic>),
          )
          .toList(growable: false);

  final String id, name, barangay, municipality;
  final double? latitude, longitude;
  final String connectionStatus, trend, source;
  final bool isOnline, isStale;
  final int? ageSeconds;
  final AlertLevel? alertLevel;
  final double? currentDepthCm, latestDepthCm, latestRainfallMm;
  final String? sensorQuality, firmwareVersion;
  final DateTime? observedAt;
  final double thresholdAdvisoryCm, thresholdWarningCm, thresholdEvacuateCm;
  final List<MonitoringSample> history;

  bool get isSimulator => source == 'simulator';
  bool get hasValidReading => currentDepthCm != null;

  static AlertLevel? _alert(String? value) => switch (value) {
    'NORMAL' => AlertLevel.normal,
    'ADVISORY' => AlertLevel.advisory,
    'WARNING' => AlertLevel.warning,
    'EVACUATE' => AlertLevel.evacuate,
    _ => null,
  };

  static DateTime? _time(dynamic value) =>
      value is String ? DateTime.tryParse(value)?.toLocal() : null;
}

class MonitoringSnapshot {
  MonitoringSnapshot.fromJson(Map<String, dynamic> json)
    : checkedAt = DateTime.parse(json['checked_at'] as String).toLocal(),
      staleAfterSeconds = json['stale_after_seconds'] as int,
      stations = (json['stations'] as List<dynamic>)
          .map(
            (value) =>
                MonitoringStation.fromJson(value as Map<String, dynamic>),
          )
          .toList(growable: false);

  final DateTime checkedAt;
  final int staleAfterSeconds;
  final List<MonitoringStation> stations;
}
